core/jvmMain/src/kotlinx/io/internal/JavaInterop.kt

```diff
@@ -0,0 +1,92 @@
+package kotlinx.io.internal
+
+import kotlinx.io.*
+import kotlinx.io.buffer.*
+import java.io.*
+
+internal class InputStreamFromInput(private val input: Input) : InputStream() {
+    override fun read(): Int {
+        if (input.exhausted()) {
+            return -1
+        }
+        return input.readByte().toInt() and 0xFF
+    }
+
+    override fun read(b: ByteArray): Int {
+        if (b.isEmpty()) return 0
+        val result = input.readAvailableTo(bufferOf(b))
+        if (result == 0) return -1
+        return result
+    }
+
+    override fun read(b: ByteArray, off: Int, len: Int): Int {
+        if (len == 0) return 0
+        val result = input.readAvailableTo(bufferOf(b), off, off + len)
+        if (result == 0) return -1
+        return result
+    }
+
+    override fun close() {
+        input.close()
+    }
+}
+
+internal class InputFromInputStream(private val inputStream: InputStream) : Input() {
+    override fun closeSource() {
+        inputStream.close()
+    }
+
+    override fun fill(buffer: Buffer, startIndex: Int, endIndex: Int): Int {
+        // Zero-copy attempt
+        if (buffer.buffer.hasArray()) {
+            return inputStream.read(buffer.buffer.array(), startIndex, endIndex - startIndex)
+                .coerceAtLeast(0) // -1 when IS is closed
```

Comment:
please consider avoiding index calculations in arguments(especially with methods or corner cases)